File: MATLAB/clipper.m
function [u] = clipper(A)
%% Inputs:
%
%       A:  an nxn unweighted, symmetric adjacency matrix. 
%% Outputs:
%
%       u:      final iterate of the gradient ascent scheme

I = eye(size(A));
M = A + I;
n = length(M); % input size

Mc = ones(size(M)) - M; % binay complement of constraint matrix M

u0 = rand(n,1);

%% Set parameters
tolu = 1e-8; 
tolF = 1e-8;  
sigma = 0.01;
beta = 0.5;  % for reduction of alpha based on backtracking line search
maxlsiters = 99; % maximum number of line search iterations
maxiniters = 200; % maximum number of gradient ascent steps for each d
maxoliters = 1000; % maximum number of outer loop iterations
epsilon = 1e-9; % numerical threshold to replace 0 with 100x machine epsilon



% One step of power method to have a good scaling of u
u = M * u0;
u = u / norm(u);

% Initial d 
mcu = Mc * u;
idxD = (mcu > epsilon) & (u > epsilon); % indices where mcu>0 and u>0
   if any(idxD)
   Mu = M * u;
   d = mean( Mu(idxD)./mcu(idxD) ) ;% smallest d that makes ?u= Mu - dmcu  hit the positive orthant boundaries (for indices where mcu>0 and u>0)
   else
   d = 0; % in this case matrix Mc is all zeros (or there%s no active constraint)
   end
   
Md = M - d * Mc;


%% Projected gradient ascent loop

i = 1;
outerloop = true;
 while outerloop && i<maxoliters
    F = u.' * Md * u; % object value
    j = 1;
    innerloop = true;
    while innerloop && j<maxiniters
        gradF = Md * u; % gradient
        gradFop = gradF - gradF.'*u * u; %
        alpha = 1;
        k = 1;
        armijo = false;
        while ~armijo && k<maxlsiters
            unew = u + alpha * gradFop;    % gradient step
            unew = max(unew/norm(unew), 0); 
            Fnew = unew.' * Md * unew;     % new objective value
            gradfopnew = 2*Md*unew -unew*unew'*2*Md*unew;
            deltau = u - unew;
            deltaF = Fnew - F;             % change in objective value
            
            if (deltaF < norm(sigma * gradFop*u'*deltau)) % armijo steps
                alpha = alpha * beta; 
            else
                alpha = alpha / beta ^ 0.5;
                armijo = true;
            end
            
            k = k + 1;
        end

        deltau = norm(unew - u); % change in vector u
        if ( (deltau < tolu) || (abs(deltaF) < tolF) ) 
            innerloop = false; % if desired accuracy reached break from inner loop
        end
        
        % update values
        F = Fnew;
        u = unew;
        
        j = j + 1;
    end
    
    % Increase d
    mcu = Mc * u;
   idxD = (mcu > epsilon) & (u > epsilon); % indices where mcu>0 and u>0 
     if any(idxD)
       Mu = M * u;
       deltad = mean(abs( Mu(idxD)./mcu(idxD) )); % Find smallest deltad that makes gradF=Mu-dmcu  hit the positive orthant boundaries (for indices where mcu>0 and u>0)
        
        d = d + deltad; % increase d
        Md = M - d * Mc; % update matrix Md
     end
       if (all(u == 0) || (numel(unique(u)) == 2 && ismember(0, unique(u))))
            outerloop = false;
       end

    i = i + 1;
end


 
end

File: cpp/Src/clipper.cpp
#include"clipper.hpp"



VectorXd clipper(const MatrixXd& A, const VectorXd& u0)
{
    int n = A.rows();
    MatrixXd M = MatrixXd::Identity(n,n) + A;
    
    MatrixXd M_C  = MatrixXd::Ones(n,n) - M ;
    MatrixXd Md (n,n);

double tolu = 1e-8; 
double tolF = 1e-8;  
double sigma = 0.01;
double beta = 0.5;  
int maxlsiters = 99; 
int maxiniters = 200; 
int maxoliters = 1000; 
double epsilon = 1e-9; 
VectorXd deltau(n);
double deltaF;
double d = 2;
double deltad = 0.5;

VectorXd u(n), Mu(n), gradF(n), gradFop(n), unew(n), mcu(n),gradfopnew(n);
VectorXd mu0(n);
mu0 = M * u0;
u = mu0 / mu0.norm();
Md = M - d * M_C; 


int i = 0;
bool outerloop = true;

while(outerloop && i < maxoliters){

 double F = u.transpose() * Md * u;
 double Fnew{};
 double alpha = 1;
 int j = 0;
 bool innerloop = true;
  while(innerloop && j < maxiniters){
     
        gradF = 2 * Md * u; // gradient
        
        gradFop = gradF - gradF.dot(u) * u;; 
         int k = 1;
         bool armijo = false;
         while (!armijo && k < maxlsiters) {
           
         
            unew = u + alpha * gradFop;    // gradient step
            unew = unew/unew.norm(); 
            unew = unew.array().max(0); // Apply element-wise maximum with 
     
            Fnew = unew.transpose() * Md * unew;     //new objective value
            
            gradfopnew = 2 * Md *unew - unew * unew.transpose() * 2 * Md * unew;

             deltau = u - unew;
             deltaF = Fnew - F;
            VectorXd armijo_step = sigma * gradFop * u.transpose() * deltau;
            double norm_armijo_step = armijo_step.norm();
             if (deltaF < norm_armijo_step){
                alpha = alpha * beta;
             }
             else{
                alpha = alpha / sqrt(beta);
                armijo = true;
             }
             k++;
            
         }
         double norm_deltau = deltau.norm();
           if ((norm_deltau < tolu) || (std::abs(deltaF) < tolF)){
                innerloop = false;
          }
         F = Fnew;
         u = unew;
         j++;
  }
     std::set<double> uniqueElements;

    //  Insert elements of the vector into the set (sets automatically store only unique elements)
    for (int i = 0; i < u.size(); ++i) {
        uniqueElements.insert(u[i]);
      }
    if (uniqueElements.size() == 2){
     outerloop = false;
      }
    d = d + deltad;
    Md = M - d * M_C; 

    i++;
}

return u;
   



}